src/librustc/traits/select.rs

```diff
@@ -690,10 +690,76 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 }
             }
 
-            ty::Predicate::TypeOutlives(..) | ty::Predicate::RegionOutlives(..) => {
-                // we do not consider region relationships when
-                // evaluating trait matches
-                Ok(EvaluatedToOk)
+            ty::Predicate::TypeOutlives(ref binder) => {
+                assert!(!binder.has_escaping_regions());
+                // Check if the type has higher-ranked regions.
+                if binder.skip_binder().0.has_escaping_regions() {
+                    // If so, this obligation is an error (for now). Eventually we should be
+                    // able to support additional cases here, like `for<'a> &'a str: 'a`.
+
+                    // NOTE: this hack is implemented in both trait fulfillment and
+                    // evaluation. If you fix it in one place, make sure you fix it
+                    // in the other.
+
+                    // We don't want to allow this sort of reasoning in intercrate
+                    // mode, for backwards-compatibility reasons.
+                    if self.intercrate.is_some() {
+                        Ok(EvaluatedToAmbig)
+                    } else {
+                        Ok(EvaluatedToErr)
+                    }
+                } else {
+                    // If the type has no late bound regions, then if we assign all
+                    // the inference variables in it to be 'static, then the type
+                    // will be 'static itself.
+                    //
+                    // Therefore, `staticize(T): 'a` holds for any `'a`, so this
+                    // obligation is fulfilled. Because evaluation works with
+                    // staticized types (yes I know this is involved with #21974),
+                    // we are 100% OK here.
+                    Ok(EvaluatedToOk)
+                }
+            }
+
+            ty::Predicate::RegionOutlives(ref binder) => {
+                let ty::OutlivesPredicate(r_a, r_b) = binder.skip_binder();
+
+                if r_a == r_b {
+                    // for<'a> 'a: 'a. OK
+                    Ok(EvaluatedToOk)
+                } else if r_a.is_late_bound() || r_b.is_late_bound() {
```

Comment:
Annoyingly, the current region solver has one extra case to consider: the case where `r_a` is `'static`. In that case, know it outlives any `r_b`. But we support this case inconsistently owing to the leak-check. Not sure how relevant that is here.

Reply:
In particular, even within the leak check, we don't generate a constraint like `'a <= 'static`, and hence the leak-check doesn't fail:

https://github.com/rust-lang/rust/blob/80e6e3e582a9eb9115900d51ae6edecf46d81f6e/src/librustc/infer/region_constraints/mod.rs#L731-L733

Reply:
The caching code that is going wrong — however — only cares about `predicate_must_hold`, so it's ok to falsely fail from time to time. Still, we might want this to be `is_late_bound() && is_late_bound()`? 

You mentioned coherence, though...

Reply:
Actually, I think the only way something like this might happen if someone writes a literal `'static: 'a` bound in their trait. In this case, evaluation would return an error, but fulfillment would be ok. 

Maybe I should add a special case for this (`r_a == 'static`)?